grpc/server.cpp

```diff
@@ -0,0 +1,45 @@
+#include "server.hpp"
+
+#include <blackhole/attribute.hpp>
+#include <grpcpp/server_builder.h>
+
+#include "library/logger.hpp"
+#include "read_job.hpp"
+#include "write_job.hpp"
+
+namespace ioremap { namespace grpc {
+
+server_t::server_t(dnet_node &node, ::grpc::ServerCompletionQueue &completion_queue)
+: node_(node)
+, completion_queue_(completion_queue) {
+}
+
+int server_t::run() {
+	new read_job_t(node_, completion_queue_, async_service_);
+	new write_job_t(node_, completion_queue_, async_service_);
+	return 0;
+}
+
+int server_t::make(std::unique_ptr<server_t> &server, std::unique_ptr<::grpc::CompletionQueue> &completion_queue,
+        const std::string &address, dnet_node &node) {
+
+	::grpc::ServerBuilder builder;
+	builder.AddListeningPort(address, ::grpc::InsecureServerCredentials());
+
+	auto server_completion_queue = builder.AddCompletionQueue();
+	server.reset(new server_t(node, *server_completion_queue));
+	completion_queue = std::move(server_completion_queue);
+
+	builder.RegisterService(&server->async_service_);
+
+	server->server_ = builder.BuildAndStart();
+	if (!server->server_) {
+		//TODO: log
+		return -EPERM;
```

Comment:
Let's discuss C/C++ transformation style. Here we're in C++ function, but used C style of error handling: returning code. But note that also exceptions can occur. And note that you don't free pointers here. One of good patterns:
```
void func(unique_ptr<C1> &ptr1, unique_ptr<C2> &ptr2) {
  auto p1 = make_unique<C1>;
  auto p2 = make_unique<C2>;
  // Some large code initialization, cat throw

  // And at the end (if you will not reach the end, pointers'll be freed):
  ptr1 = std::move(p1); // No throw here
  ptr1 = std::move(p2); // No throw here
```
Also, we can prefer exception error-passing style in C++ (not always, but only for errors that break the process). And we can use try-catch guard in C++ functions that could be called from C.

Reply:
`Grpc` library doesn't throw exception except `bad_alloc`. I used it for this code, but I have not yet resolved possible exceptions in all files. We can discuss possible solutions.

Reply:
f24a094